features/netsocket/NetworkInterface.h

```diff
@@ -126,6 +130,26 @@ class NetworkInterface {
      */
     virtual nsapi_error_t add_dns_server(const SocketAddress &address);
 
+    /** Register callback for status reporting
+     *
+     *  @param status_cb The callback for status changes
+     */
+    virtual void attach(mbed::Callback<void(nsapi_event_t, intptr_t)> status_cb);
```

Comment:
Could you document what are the parameters of the callback in input ?

From the signature it looks like it may handle more than just status change otherwise why not just take in a `nsapi_connection_status_t` as the callback parameter ?

Reply:
Some explanation is in nsapi_types.h, with nsapi_event_t. Is that clear enough that you understand it? (I know what it means but that may be because I was in the review meetings).

Agree it should be summarised here.

Reply:
> Is that clear enough that you understand it?

Looking at the documentation and signature; what is unclear for me is whether this status_cb will solely handle status change events or if it will receive other kind of events if nsapi_event_t is updated. 

If the parameter of the attach function is not destined to handle events other than connection status change events then the interface is overly complex.

If the callback attached handles more than status events in the future then it is a breaking change: the contract with application code stated in comments would have been broken.

Reply:
Would this suffice for the documentation:
```
    /** Register callback for status reporting
     *  
     *  The specified status callback function will be called on status changes
     *  on the network. The parameters on the callback are the event type and
     *  event-type dependent reason parameter.
     *
     *  @param status_cb The callback for status changes
     */
    virtual void attach(mbed::Callback<void(nsapi_event_t, intptr_t)> status_cb);
```

Reply:
The intent is that new events will be added to `nsapi_event_t`, and users would be expected to ignore event types they didn't know about. That should be documented. There is one new event type already being plotted.

The use of an `enum` for a set that may be extended is a little iffy; it seems to be a C++-ism to use them instead of `#define`s, despite that. Precedent has already been set by `nsapi_error` and `nsapi_socket_option_t`, which are gradually being extended.

Reply:
@pan- Are you happy with the proposed change in my comment or does it need some refinement?

Reply:
Looks good to me